flutter_mobx/lib/src/observer.dart

```diff
@@ -13,15 +22,54 @@ class Observer extends StatelessObserverWidget
     implements
         Builder {
   // ignore: prefer_const_constructors_in_immutables
-  Observer({Key key, @required this.builder})
+  Observer({Key key, @required this.builder, String name})
       : assert(builder != null),
-        super(key: key);
+        debugConstructingStackFrame = debugFindConstructingStackFrame(),
+        super(key: key, name: name);
 
   @override
   final WidgetBuilder builder;
 
+  /// The stack frame pointing to the source that constructed this instance.
+  final String debugConstructingStackFrame;
+
+  @override
+  String getName() =>
+      super.getName() +
+      (debugConstructingStackFrame != null
+          ? '\n$debugConstructingStackFrame'
+          : '');
+
   @override
   Widget build(BuildContext context) => builder(context);
+
+  /// Matches constructor stack frames, in both VM and web environments.
+  static final _constructorStackFramePattern = RegExp(r'\bnew\b');
+
+  /// Finds the first non-constructor frame in the stack trace.
+  ///
+  /// [stackTrace] defaults to [StackTrace.current].
+  @visibleForTesting
+  static String debugFindConstructingStackFrame([StackTrace stackTrace]) {
+    String stackFrame;
+
+    assert(() {
+      if (debugAddStackTraceInObserverName) {
+        final stackTraceString = (stackTrace ?? StackTrace.current).toString();
+        stackFrame = LineSplitter.split(stackTraceString)
+            // Evalulate frames starting from below this closure, method, and
+            // Observer constructor
+            .skip(3)
```

Comment:
What is the logic for skipping three levels. I think we should document it with some comments.

Reply:
Also how will this work if someone decides to use the `StatelessObserverWidget` or `StatefulObserverWidget` ?

Reply:
The comment describing the skip is right above the line. How should I clarify?

Reply:
I made a small change to the comment. That make more sense?

Reply:
Regarding the observer widgets, it doesn't — the `const` constructors prevent it.

Reply:
That, or we introduce a non-const constructor to each of those classes, which will populate a stack frame variable. Your call.

Reply:
> That, or we introduce a non-const constructor to each of those classes, which will populate a stack frame variable. Your call.

Ya we should ensure a user of the Subclass gets the right location as well. I think we should add the constructors to them.

Reply:
> The comment describing the skip is right above the line. How should I clarify?

I saw that. What I was curious was about the magic number **3**. Why is that relevant and can we qualify its use here ? As in why are we skipping 3 levels and not 4. Is there a hidden assumption here?

Reply:
1. the anonymous function in the assert
2. the `debugFindConstructingStackFrame` method
3. the constructor invoking that `debugFindConstructingStackFrame`
4. either user source (which is what we want), or an Observer subclass' constructor (which we skip past with the regex)

Reply:
> > That, or we introduce a non-const constructor to each of those classes, which will populate a stack frame variable. Your call.
> 
> Ya we should ensure a user of the Subclass gets the right location as well. I think we should add the constructors to them.

ok. i'll take a look at hoisting it tomorrow.

Reply:
> 1. the anonymous function in the assert
> 2. the `debugFindConstructingStackFrame` method
> 3. the constructor invoking `debugFindConstructingStackFrame`
> 4. either user source (which is what we want), or an Observer subclass' constructor (which we skip past with the regex)
> 
> Maybe I should include this rationale in the comment? To be very clear?

Yup! Right now we have lots of mental context, so we find this discovery trivial, but in few weeks, I am sure we will be scratching our heads again :-) as to why we skipped this many levels...

Reply:
Well, I mean, the comment already says that. Maybe just not clear enough. I'll amend.

Reply:
OK. Done.